practices/multiple_environments.md

```diff
@@ -0,0 +1,125 @@
+  - load testing
+  - devops
+  - git
+  - testing
+  - testing environment
+  - staging environment
+  - production environment
+  - continuous integration
+---
+
+# Multiple Environments
+
+**TL;DR**
+
+Multi-environment usage is a recommended practice for software development as it helps in maintaining an optimal workflow for the project which overall helps in delivering the project in a reliable and timely manner. This structured release management process allows rollout, testing, and rollback in case of problems.
+
+## What Is a Multi Environments
+
+Multi-environment is a recommended practice that allows you to ship code to the end-user without any bugs or issues. This type of practice is very reliable as it involves multiple stages or environments. It also helps the company to reply to the user's feedback efficiently and immediately.
+
+These environments are divided according to the tiers of the project. A particular individual related to a specific environment may not have access to other environments to avoid security breaches or human errors.
+
+We can classify Multi-environments into different stages, according to the level of development across environments:
+
+1. **Development environment**
+   The development environment (dev) is the environment in which software changes are developed, most simply an individual developer's workstation. This environment includes development tools like a compiler, integrated development environment, different or additional versions of libraries and support software, etc., which are not present in a user's environment.
+   A developer working on this environment has a working copy of source code on their machine, and changes are submitted to the repository, being committed either to the trunk or a branch, depending on development methodology. The environment on an individual workstation, in which changes are worked on and tried out, may be referred to as the local environment or a sandbox.
+2. **Testing environment**
+   The purpose of the test environment is to allow human testers to exercise new and changed code via either automated checks or non-automated techniques. After the developer accepts the new code and configurations through unit testing in the development environment, the items are moved to one or more test environments. Upon test failure, the test environment can remove the faulty code from the test platforms, contact the responsible developer, and provide detailed test and result logs. 
+   If all tests pass, the test environment or a continuous integration framework controlling the tests can automatically promote the code to the next deployment environment.
+3. **Staging environment**
+   The staging environment or pre-production environment is an environment for testing that exactly resembles a production environment. It seeks to mirror an actual production environment as closely as possible and may connect to other production services and data, such as databases. 
```

Comment:
Please, this senction too closely follows the original text from Wiki:

_A stage, staging or pre-production environment is an environment for testing that exactly resembles a production environment.[7] It seeks to mirror an actual production environment as closely as possible and may connect to other production services and data, such as databases. For example, servers will be run on remote machines, rather than locally (as on a developer's workstation during dev, or on a single test machine during test), which tests the effects of networking on the system._

_The primary use of a staging environment is to test all the installation/configuration/migration scripts and procedures before they're applied to a production environment. This ensures all major and minor upgrades to a production environment are completed reliably, without errors, and in a minimum of time._

_Another important use of staging is performance testing, particularly load testing, as this is often sensitive to the environment._

_Staging is also used by some organizations to preview new features to select customers or to validate integrations with live versions of external dependencies._